File: EclipseWorkspace/ScalaServer/src/perlinnoise.scala
object PerlinNoise
{
    def GenerateWhiteNoise(width: Int, height: Int): Float[][] =
		{
      Float[][] noise = Misc.GetEmptyArray<Float>(width, height)

			for (Int i = 0; i < width; i++)
			{
				for (Int j = 0; j < height; j++)
				{
					noise[i][j] = (Float)Settings.Random.NextDouble()
				}
			}
			return noise
		}

		def Interpolate(x0: Float, x1: Float, alpha: Float): Float =
		{
			return x0 * (1 - alpha) + alpha * x1
		}

		def Interpolate(minY: Int, maxY: Int, t: Float): Int =
		{
			Float u = 1 - t
			return (Int)(minY * u + maxY * t)
		}

		def MapInts(minY: Int, maxY: Int, perlinNoise: Float[][]): Int[][] =
		{
			Int width = perlinNoise.Length
			Int height = perlinNoise[0].Length
      Int[][] heightMap = Misc.GetEmptyArray<Int>(width, height)

			for (Int i = 0; i < width; i++)
			{
				for (Int j = 0; j < height; j++)
				{
					heightMap[i][j] = Interpolate(minY, maxY, perlinNoise[i][j])
				}
			}
			return heightMap
		}

		def MapFloats(minY: Float, maxY: Float, perlinNoise: Float[][]): Float[][] =
		{
			Int width = perlinNoise.Length
			Int height = perlinNoise[0].Length
			Float[][] treeMap = Misc.GetEmptyArray<Float>(width, height)

			for (Int i = 0; i < width; i++)
			{
				for (Int j = 0; j < height; j++)
				{
					treeMap[i][j] = Interpolate(minY, maxY, perlinNoise[i][j])
				}
			}
			return treeMap
		}

		def GenerateSmoothNoise(baseNoise: Float[][], octave: Int): Float[][] =
		{
			Int width = baseNoise.Length
			Int height = baseNoise[0].Length

      Float[][] smoothNoise = Misc.GetEmptyArray<Float>(width, height)
			Int samplePeriod = 1 << octave // calculates 2 ^ k
			Float sampleFrequency = 1.0f / samplePeriod

			for (Int i = 0; i < width; i++)
			{
				//calculate the horizontal sampling indices
				Int iSample0 = (i / samplePeriod) * samplePeriod
				Int iSample1 = (iSample0 + samplePeriod) % width //wrap around
				Float horizontalBlend = (i - iSample0) * sampleFrequency

				for (Int j = 0; j < height; j++)
				{
					//calculate the vertical sampling indices
					Int jSample0 = (j / samplePeriod) * samplePeriod
					Int jSample1 = (jSample0 + samplePeriod) % height //wrap around
					Float verticalBlend = (j - jSample0) * sampleFrequency

					//blend the top two corners
					Float top = Interpolate(baseNoise[iSample0][jSample0], baseNoise[iSample1][jSample0], horizontalBlend)

					//blend the bottom two corners
					Float bottom = Interpolate(baseNoise[iSample0][jSample1], baseNoise[iSample1][jSample1], horizontalBlend)

					//final blend
					smoothNoise[i][j] = Interpolate(top, bottom, verticalBlend)
				}
			}
			return smoothNoise
		}

		def GeneratePerlinNoise(baseNoise: Float[][], octaveCount: Int): Float[][] =
		{
			Int width = baseNoise.Length
			Int height = baseNoise[0].Length

			var smoothNoise = new Float[octaveCount][][] //an array of 2D arrays containing

			const Float PERSISTANCE = 0.4f

			//generate smooth noise
			for (Int i = 0; i < octaveCount; i++)
			{
				smoothNoise[i] = GenerateSmoothNoise(baseNoise, i)
			}

      Float[][] perlinNoise = Misc.GetEmptyArray<Float>(width, height) //an array of floats initialised to 0
			Float amplitude = 1f
			Float totalAmplitude = 0.0f

			//blend noise together
			for (Int octave = octaveCount - 1; octave >= 0; octave--)
			{
				amplitude *= PERSISTANCE
				totalAmplitude += amplitude

				for (Int i = 0; i < width; i++)
				{
					for (Int j = 0; j < height; j++)
					{
						perlinNoise[i][j] += smoothNoise[octave][i][j] * amplitude
					}
				}
			}

			//normalisation
			for (Int i = 0; i < width; i++)
			{
				for (Int j = 0; j < height; j++)
				{
					perlinNoise[i][j] /= totalAmplitude
				}
			}
			return perlinNoise
		}

		def GetIntMap(width :Int, height: Int, minY: Int, maxY: Int, octaveCount: Int): Int[][] =
		{
      Float[][] baseNoise = GenerateWhiteNoise(width, height)
			Float[][] perlinNoise = GeneratePerlinNoise(baseNoise, octaveCount)
			return MapInts(minY, maxY, perlinNoise)
		}

    def GetFloatMap(width: Int, height: Int, minY: Float, maxY: Float, octaveCount: Int): Float[][] =
		{
      Float[][] baseNoise = GenerateWhiteNoise(width, height)
			Float[][] perlinNoise = GeneratePerlinNoise(baseNoise, octaveCount)
			return MapFloats(minY, maxY, perlinNoise)
		}
}

File: DotNet/WorldServer/World/Generator/PerlinNoise.cs
using System;

namespace Sean.World
{
    // see http://flafla2.github.io/2014/08/09/perlinnoise.html

	internal class PerlinNoise
	{


        private static Array<float> GenerateSmoothNoise(Array<float> baseNoise, int octave)
		{
            var smoothNoise = new Array<float> (baseNoise.Size);
            int samplePeriod = 1 << (octave + baseNoise.Size.scale); // calculates 2 ^ k
			float sampleFrequency = 1.0f / samplePeriod;

            for (int i = baseNoise.Size.minZ; i < baseNoise.Size.maxZ; i += baseNoise.Size.scale) 
			{
				//calculate the horizontal sampling indices
				int iSample0 = (i / samplePeriod) * samplePeriod;
                int iSample1 = (iSample0 + samplePeriod);
                if (iSample1 > baseNoise.Size.maxZ) iSample1 = iSample1 - baseNoise.Size.zWidth; //wrap around
				float horizontalBlend = (i - iSample0) * sampleFrequency;

                for (int j = baseNoise.Size.minX; j < baseNoise.Size.maxX; j += baseNoise.Size.scale) 
				{
					//calculate the vertical sampling indices
					int jSample0 = (j / samplePeriod) * samplePeriod;
                    int jSample1 = (jSample0 + samplePeriod);
                    if (jSample1 > baseNoise.Size.maxX) jSample1 = jSample1 - baseNoise.Size.xHeight; //wrap around
					float verticalBlend = (j - jSample0) * sampleFrequency;

					//blend the top two corners
                    float top = Lerp(baseNoise.Get(jSample0,iSample0), baseNoise.Get(jSample0,iSample1), horizontalBlend);

					//blend the bottom two corners
                    float bottom = Lerp(baseNoise.Get(jSample1,iSample0), baseNoise.Get(jSample1,iSample1), horizontalBlend);

					//final blend
                    smoothNoise.Set(j,i,Lerp(top, bottom, verticalBlend));
				}
			}
			return smoothNoise;
		}

        public static Array<int> GetIntMap(ArraySize size, int minX, int maxX, int minZ, int maxZ, int octaveCount)
        {
            var perlin = new PerlinNoise();
            var noise = new Array<int>(size);
            for (int z = minZ; z < maxZ; z += noise.Size.scale)
            {
                for (int x = minX; x < maxX; x += noise.Size.scale)
                {
                    double height = perlin.OctavePerlin (noise.Size, x,1,z, octaveCount, 1.0);
                    noise.Set(x,z, (int)(height*10));
                }
            }
            return noise;
        }

        public static Array<float> GetFloatMap(ArraySize size, int minX, int maxX, int minZ, int maxZ, int octaveCount)
        {
            var perlin = new PerlinNoise();
            var noise = new Array<float>(size);
            for (int z = minZ; z < maxZ; z += noise.Size.scale)
            {
                for (int x = minX; x < maxX; x += noise.Size.scale)
                {
                    double height = perlin.OctavePerlin(noise.Size, x, 1, z, octaveCount, 1.0);
                    noise.Set(x, z, (float)(height * 10));
                }
            }
            return noise;
        }

        public double OctavePerlin(ArraySize size, int x, int y, int z, int octaves, double persistence) {
            double xf = (double)x / size.maxX;
            double yf = 0.0;
            double zf = (double)z / size.maxZ;
            double total = 0;
            int frequency = 1;
            double amplitude = 1;
            double maxValue = 0;            // Used for normalizing result to 0.0 - 1.0
            for(int i=0;i<octaves;i++) {
                total += Perlin(size, xf, yf, zf) * amplitude;

                maxValue += amplitude;

                amplitude *= persistence;
                frequency *= 2;
            }

            return total/maxValue;
        }

        private int p(int x, int y, int z) {
            return (int)(Misc.GetDeterministicInt (x, y, z, WorldSeed) % 256);
        }

        public double Perlin(ArraySize size, double x, double y, double z) {
            //if(repeat > 0) {                                    // If we have any repeat on, change the coordinates to their "local" repetitions
            //    x = x%repeat;
            //    y = y%repeat;
            //    z = z%repeat;
            //}

            int xi = (int)x;// & 255;                     // Calculate the "unit cube" that the point asked will be located in
            int yi = (int)y;// & 255;                     // The left bound is ( |_x_|,|_y_|,|_z_| ) and the right bound is that
            int zi = (int)z;// & 255;                     // plus 1.  Next we calculate the location (from 0.0 to 1.0) in that cube.
            double xf = x-(int)x;                         // We also fade the location to smooth the result.
            double yf = y-(int)y;
            double zf = z-(int)z;

            int aaa, aba, aab, abb, baa, bba, bab, bbb;
            aaa = p(    xi ,    yi ,    zi );
            aba = p(    xi , (yi++),    zi );
            aab = p(    xi ,    yi , (zi++));
            abb = p(    xi , (yi++), (zi++));
            baa = p( (xi++),    yi ,    zi );
            bba = p( (xi++), (yi++),    zi );
            bab = p( (xi++),    yi , (zi++));
            bbb = p( (xi++), (yi++), (zi++));

            double u = xf;//fade(xf);
            double v = yf;//fade(yf);
            double w = zf;//fade(zf);

            double x1, x2, y1, y2;
            x1 = Lerp(    Grad (aaa, xf  , yf  , zf),   // The gradient function calculates the dot product between a pseudorandom
                Grad (baa, xf-1, yf  , zf),             // gradient vector and the vector from the input coordinate to the 8
                u);                                     // surrounding points in its unit cube.
            x2 = Lerp(    Grad (aba, xf  , yf-1, zf),   // This is all then lerped together as a sort of weighted average based on the faded (u,v,w)
                Grad (bba, xf-1, yf-1, zf),             // values we made earlier.
                u);
            y1 = Lerp(x1, x2, v);

            x1 = Lerp(    Grad (aab, xf  , yf  , zf-1),
                Grad (bab, xf-1, yf  , zf-1),
                u);
            x2 = Lerp(    Grad (abb, xf  , yf-1, zf-1),
                Grad (bbb, xf-1, yf-1, zf-1),
                u);
            y2 = Lerp (x1, x2, v);

            return (Lerp (y1, y2, w)+1)/2;               // For convenience we bind the result to 0 - 1 (theoretical min/max before is [-1, 1])
        }

        public static double Grad(int hash, double x, double y, double z) {
            switch(hash & 0xF)
            {
            case 0x0: return  x + y;
            case 0x1: return -x + y;
            case 0x2: return  x - y;
            case 0x3: return -x - y;
            case 0x4: return  x + z;
            case 0x5: return -x + z;
            case 0x6: return  x - z;
            case 0x7: return -x - z;
            case 0x8: return  y + z;
            case 0x9: return -y + z;
            case 0xA: return  y - z;
            case 0xB: return -y - z;
            case 0xC: return  y + x;
            case 0xD: return -y + z;
            case 0xE: return  y - x;
            case 0xF: return -y - z;
            default: return 0; // never happens
            }
        }

        private static float Lerp(float a, float b, float w)
        {
            return a + w * (b - a);
        }
        public static double Lerp(double a, double b, double w)
        {
            return a + w * (b - a);
        }

        private static int Lerp(int minY, int maxY, float t)
        {
            float u = 1 - t;
            return (int)(minY * u + maxY * t);
        }

        private static double Fade(double t)
        {
            // Fade function as defined by Ken Perlin.  This eases coordinate values
            // so that they will "ease" towards integral values.  This ends up smoothing
            // the final output.
            return t * t * t * (t * (t * 6 - 15) + 10);         // 6t^5 - 15t^4 + 10t^3
        }

        public int WorldSeed { get; set; }
        //public int repeat;
	}
}